## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :x: [more...](#cffe4793-67d8-4163-bf1e-0192f4996fb6) |
| mint-large-bucket.sh | :x: [more...](#7f6eee14-e64c-4bdd-b573-83c5c9893c64) |
| mint-fs.sh | :x: [more...](#3d9badf6-082a-4b0a-8135-6fe5d95cb973) |
| mint-dist-xl.sh | :x: [more...](#e4eb74bf-3664-493c-904e-cd54396e394d) |
| mint-gateway-s3.sh | :x: [more...](#97413d58-5366-4c02-a624-083a27ef4d4e) |
| mint-gateway-azure.sh | :x: [more...](#632d32c6-9920-4c8c-a52a-21059518366f) |
| mint-gateway-nas.sh | :x: [more...](#384f5edc-18c6-4d96-b8e9-33fa8566189c) |
<a name="632d32c6-9920-4c8c-a52a-21059518366f"></a>
### 9697-9fb7779/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31359
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bfea69b1e59f:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 2 minutes and 40 seconds
{
  "name": "aws-sdk-php",
  "function": "headObject ( array $params = [] )",
  "args": {
    "aws-sdk-php-5ecb1d60c80f9": "obj1",
    "aws-sdk-php-5ecb1d60c80fa": "obj2"
  },
  "duration": "81",
  "status": "FAIL",
  "error": "headObject API Metadata didn't match for aws-sdk-php-5ecb1d60c80f9/obj1"
}
(3/15) Running aws-sdk-ruby tests ... done in 30 seconds
(4/15) Running awscli tests ... done in 2 minutes and 56 seconds
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... done in 2 minutes and 59 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 21 seconds
(8/15) Running minio-go tests ... done in 3 minutes and 40 seconds
(9/15) Running minio-java tests ... done in 25 minutes and 42 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 34 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 46 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 25 seconds
(13/15) Running s3select tests ... done in 1 minutes and 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="3d9badf6-082a-4b0a-8135-6fe5d95cb973"></a>
### 9697-9fb7779/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31729
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c8dd0397467b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "14",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:27.082623854Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 1 seconds
(4/15) Running awscli tests ... done in 2 minutes and 5 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 42 seconds
(7/15) Running minio-dotnet tests ... done in 32 seconds
(8/15) Running minio-go tests ... done in 41 seconds
(9/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 34,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:27.082623854Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 46 seconds
(10/15) Running minio-py tests ... FAILED in 1 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 24,
  "message": "time data '2020-05-25T01:20:27.082623854Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:27.082623854Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:27.082623854Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:27.082623854Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(10/15) Running s3cmd tests ... done in 23 seconds
(11/15) Running s3select tests ... done in 2 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="cffe4793-67d8-4163-bf1e-0192f4996fb6"></a>
### 9697-9fb7779/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31967
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e8e99a319b91:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "15",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:26.278469907Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 2 seconds
(4/15) Running awscli tests ... done in 2 minutes and 11 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 48 seconds
(7/15) Running minio-dotnet tests ... done in 30 seconds
(8/15) Running minio-go tests ... done in 53 seconds
(9/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 47,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:26.278469907Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... FAILED in 20 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 4,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(9/15) Running minio-py tests ... FAILED in 0 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 23,
  "message": "time data '2020-05-25T01:20:26.278469907Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.278469907Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.278469907Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:26.278469907Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(9/15) Running s3cmd tests ... done in 29 seconds
(10/15) Running s3select tests ... done in 3 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="7f6eee14-e64c-4bdd-b573-83c5c9893c64"></a>
### 9697-9fb7779/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31561
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 76a82b35b2e5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 4 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "16",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:31.536816465Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 8 seconds
(4/15) Running awscli tests ... done in 2 minutes and 31 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 49 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 12 seconds
(8/15) Running minio-go tests ... done in 2 minutes and 0 seconds
(9/15) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 259,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:31.536816465Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 1 minutes and 0 seconds
(10/15) Running minio-py tests ... FAILED in 1 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 263,
  "message": "time data '2020-05-25T01:20:31.536816465Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:31.536816465Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:31.536816465Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:31.536816465Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(10/15) Running s3cmd tests ... done in 37 seconds
(11/15) Running s3select tests ... done in 22 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="e4eb74bf-3664-493c-904e-cd54396e394d"></a>
### 9697-9fb7779/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31764
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fdd21525ca3e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "15",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:27.980882184Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 3 seconds
(4/15) Running awscli tests ... done in 2 minutes and 27 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 1 minutes and 0 seconds
(7/15) Running minio-dotnet tests ... done in 49 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 27 seconds
(9/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 108,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:27.980882184Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 50 seconds
(10/15) Running minio-py tests ... FAILED in 2 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 108,
  "message": "time data '2020-05-25T01:20:27.980882184Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:27.980882184Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:27.980882184Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:27.980882184Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(10/15) Running s3cmd tests ... done in 29 seconds
(11/15) Running s3select tests ... done in 9 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="384f5edc-18c6-4d96-b8e9-33fa8566189c"></a>
### 9697-9fb7779/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30859
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 11023cfa4c8c:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "15",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:26.765505431Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 2 seconds
(4/15) Running awscli tests ... done in 2 minutes and 5 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 48 seconds
(7/15) Running minio-dotnet tests ... done in 32 seconds
(8/15) Running minio-go tests ... done in 38 seconds
(9/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 44,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:26.765505431Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... FAILED in 15 seconds
{
  "name": "minio-js",
  "function": "getBucketPolicy(bucketName, cb)",
  "args": "bucketName:minio-js-test-fe226a8b-9dec-4b06-85bb-18f3144dc9c1",
  "duration": 11,
  "status": "FAIL",
  "error": "S3Error: The bucket policy does not exist at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(9/15) Running minio-py tests ... FAILED in 0 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 18,
  "message": "time data '2020-05-25T01:20:26.765505431Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.765505431Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.765505431Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:26.765505431Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(9/15) Running s3cmd tests ... done in 16 seconds
(10/15) Running s3select tests ... done in 2 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="97413d58-5366-4c02-a624-083a27ef4d4e"></a>
### 9697-9fb7779/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30424
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1df736d74855:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "name": "aws-sdk-php",
  "function": "listBuckets ( array $params = [] )",
  "args": [],
  "duration": "19",
  "status": "FAIL",
  "error": "DateTime::__construct(): Failed to parse time string (2020-05-25T01:20:26.684957464Z) at position 0 (2): The timezone could not be found in the database"
}
(3/15) Running aws-sdk-ruby tests ... done in 2 seconds
(4/15) Running awscli tests ... done in 2 minutes and 3 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 39 seconds
(7/15) Running minio-dotnet tests ... done in 30 seconds
(8/15) Running minio-go tests ... done in 42 seconds
(9/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 33,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2020-05-25T01:20:26.684957464Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioClient.listBuckets(MinioClient.java:3126), FunctionalTest.listBuckets_test(FunctionalTest.java:325), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 36 seconds
(10/15) Running minio-py tests ... FAILED in 0 seconds
{
  "name": "minio-py:test_list_buckets",
  "function": "list_buckets()",
  "duration": 24,
  "message": "time data '2020-05-25T01:20:26.684957464Z' does not match format '%Y-%m-%dT%H:%M:%SZ'",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 775, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%S.%fZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.684957464Z' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 272, in test_list_buckets\n    buckets = client.list_buckets()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 415, in list_buckets\n    return parse_list_buckets(response.data)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 198, in parse_list_buckets\n    for buckets in root.findall('Buckets')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 199, in <listcomp>\n    for bucket in buckets.findall('Bucket')\n  File \"/usr/local/lib/python3.6/dist-packages/minio/parsers.py\", line 136, in get_localized_time_elem\n    return _iso8601_to_utc_datetime(self.get_child_text(name))\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 777, in _iso8601_to_utc_datetime\n    parsed_date = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')\n  File \"/usr/lib/python3.6/_strptime.py\", line 565, in _strptime_datetime\n    tt, fraction = _strptime(data_string, format)\n  File \"/usr/lib/python3.6/_strptime.py\", line 362, in _strptime\n    (data_string, format))\nValueError: time data '2020-05-25T01:20:26.684957464Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1979, in main\n    test_list_buckets(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 279, in test_list_buckets\n    raise Exception(err)\nException: time data '2020-05-25T01:20:26.684957464Z' does not match format '%Y-%m-%dT%H:%M:%SZ'\n",
  "status": "FAIL"
}
(10/15) Running s3cmd tests ... done in 1 minutes and 4 seconds
(11/15) Running s3select tests ... done in 2 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/admin-heal-ops.go
+++ b/cmd/admin-heal-ops.go
@@ -221,8 +221,7 @@ func (ahs *allHealState) LaunchNewHealSequence(h *healSequence) (
 	// Check if new heal sequence to be started overlaps with any
 	// existing, running sequence
 	for k, hSeq := range ahs.healSeqMap {
-		if !hSeq.hasEnded() && (strings.HasPrefix(k, h.path) ||
-			strings.HasPrefix(h.path, k)) {
+		if !hSeq.hasEnded() && (HasPrefix(k, h.path) || HasPrefix(h.path, k)) {
 
 			errMsg = "The provided heal sequence path overlaps with an existing " +
 				fmt.Sprintf("heal path: %s", k)

--- a/cmd/api-headers.go
+++ b/cmd/api-headers.go
@@ -83,7 +83,7 @@ func setObjectHeaders(w http.ResponseWriter, objInfo ObjectInfo, rs *HTTPRangeSp
 	}
 
 	if strings.Contains(objInfo.ETag, "-") && len(objInfo.Parts) > 0 {
-		w.Header().Set(xhttp.AmzMpPartsCount, strconv.Itoa(len(objInfo.Parts)))
+		w.Header()[xhttp.AmzMpPartsCount] = []string{strconv.Itoa(len(objInfo.Parts))}
 	}
 
 	if objInfo.ContentType != "" {
@@ -106,18 +106,29 @@ func setObjectHeaders(w http.ResponseWriter, objInfo ObjectInfo, rs *HTTPRangeSp
 	// Set tag count if object has tags
 	tags, _ := url.ParseQuery(objInfo.UserTags)
 	tagCount := len(tags)
-	if tagCount != 0 {
-		w.Header().Set(xhttp.AmzTagCount, strconv.Itoa(tagCount))
+	if tagCount > 0 {
+		w.Header()[xhttp.AmzTagCount] = []string{strconv.Itoa(tagCount)}
 	}
 
 	// Set all other user defined metadata.
 	for k, v := range objInfo.UserDefined {
-		if HasPrefix(k, ReservedMetadataPrefix) {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) {
 			// Do not need to send any internal metadata
 			// values to client.
 			continue
 		}
-		w.Header().Set(k, v)
+		var isSet bool
+		for _, userMetadataPrefix := range userMetadataKeyPrefixes {
+			if !strings.HasPrefix(k, userMetadataPrefix) {
+				continue
+			}
+			w.Header()[strings.ToLower(k)] = []string{v}
+			isSet = true
+			break
+		}
+		if !isSet {
+			w.Header().Set(k, v)
+		}
 	}
 
 	totalObjectSize, err := objInfo.GetActualSize()

--- a/cmd/api-response.go
+++ b/cmd/api-response.go
@@ -33,8 +33,10 @@ import (
 )
 
 const (
-	timeFormatAMZLong = "2006-01-02T15:04:05.000Z" // Reply date format with nanosecond precision.
-	maxObjectList     = 10000                      // Limit number of objects in a listObjectsResponse.
+	// RFC3339 a subset of the ISO8601 timestamp format. e.g 2014-04-29T18:30:38Z
+	iso8601TimeFormat = "2006-01-02T15:04:05.000Z" // Reply date format with nanosecond precision.
+	maxObjectList     = 50000                      // Limit number of objects in a listObjectsResponse/listObjectsVersionsResponse.
+	maxDeleteList     = 10000                      // Limit number of objects deleted in a delete call.
 	maxUploadsList    = 10000                      // Limit number of uploads in a listUploadsResponse.
 	maxPartsList      = 10000                      // Limit number of parts in a listPartsResponse.
 )
@@ -400,7 +402,7 @@ func generateListBucketsResponse(buckets []BucketInfo) ListBucketsResponse {
 	for _, bucket := range buckets {
 		var listbucket = Bucket{}
 		listbucket.Name = bucket.Name
-		listbucket.CreationDate = bucket.Created.UTC().Format(timeFormatAMZLong)
+		listbucket.CreationDate = bucket.Created.UTC().Format(iso8601TimeFormat)
 		listbuckets = append(listbuckets, listbucket)
 	}
 
@@ -424,7 +426,7 @@ func generateListVersionsResponse(bucket, prefix, marker, delimiter, encodingTyp
 			continue
 		}
 		content.Key = s3EncodeName(object.Name, encodingType)
-		content.LastModified = object.ModTime.UTC().Format(timeFormatAMZLong)
+		content.LastModified = object.ModTime.UTC().Format(iso8601TimeFormat)
 		if object.ETag != "" {
 			content.ETag = "\"" + object.ETag + "\""
 		}
@@ -440,9 +442,9 @@ func generateListVersionsResponse(bucket, prefix, marker, delimiter, encodingTyp
 		content.IsLatest = true
 		versions = append(versions, content)
 	}
+
 	data.Name = bucket
 	data.Versions = versions
-
 	data.EncodingType = encodingType
 	data.Prefix = s3EncodeName(prefix, encodingType)
 	data.KeyMarker = s3EncodeName(marker, encodingType)
@@ -475,7 +477,7 @@ func generateListObjectsV1Response(bucket, prefix, marker, delimiter, encodingTy
 			continue
 		}
 		content.Key = s3EncodeName(object.Name, encodingType)
-		content.LastModified = object.ModTime.UTC().Format(timeFormatAMZLong)
+		content.LastModified = object.ModTime.UTC().Format(iso8601TimeFormat)
 		if object.ETag != "" {
 			content.ETag = "\"" + object.ETag + "\""
 		}
@@ -525,7 +527,7 @@ func generateListObjectsV2Response(bucket, prefix, token, nextToken, startAfter,
 			continue
 		}
 		content.Key = s3EncodeName(object.Name, encodingType)
-		content.LastModified = object.ModTime.UTC().Format(timeFormatAMZLong)
+		content.LastModified = object.ModTime.UTC().Format(iso8601TimeFormat)
 		if object.ETag != "" {
 			content.ETag = "\"" + object.ETag + "\""
 		}
@@ -539,7 +541,7 @@ func generateListObjectsV2Response(bucket, prefix, token, nextToken, startAfter,
 		if metadata {
 			content.UserMetadata = make(StringMap)
 			for k, v := range CleanMinioInternalMetadataKeys(object.UserDefined) {
-				if HasPrefix(k, ReservedMetadataPrefix) {
+				if strings.HasPrefix(k, ReservedMetadataPrefix) {
 					// Do not need to send any internal metadata
 					// values to client.
 					continue
@@ -574,7 +576,7 @@ func generateListObjectsV2Response(bucket, prefix, token, nextToken, startAfter,
 func generateCopyObjectResponse(etag string, lastModified time.Time) CopyObjectResponse {
 	return CopyObjectResponse{
 		ETag:         "\"" + etag + "\"",
-		LastModified: lastModified.UTC().Format(timeFormatAMZLong),
+		LastModified: lastModified.UTC().Format(iso8601TimeFormat),
 	}
 }
 
@@ -582,7 +584,7 @@ func generateCopyObjectResponse(etag string, lastModified time.Time) CopyObjectR
 func generateCopyObjectPartResponse(etag string, lastModified time.Time) CopyObjectPartResponse {
 	return CopyObjectPartResponse{
 		ETag:         "\"" + etag + "\"",
-		LastModified: lastModified.UTC().Format(timeFormatAMZLong),
+		LastModified: lastModified.UTC().Format(iso8601TimeFormat),
 	}
 }
 
@@ -627,7 +629,7 @@ func generateListPartsResponse(partsInfo ListPartsInfo, encodingType string) Lis
 		newPart.PartNumber = part.PartNumber
 		newPart.ETag = "\"" + part.ETag + "\""
 		newPart.Size = part.Size
-		newPart.LastModified = part.LastModified.UTC().Format(timeFormatAMZLong)
+		newPart.LastModified = part.LastModified.UTC().Format(iso8601TimeFormat)
 		listPartsResponse.Parts[index] = newPart
 	}
 	return listPartsResponse
@@ -657,7 +659,7 @@ func generateListMultipartUploadsResponse(bucket string, multipartsInfo ListMult
 		newUpload := Upload{}
 		newUpload.UploadID = upload.UploadID
 		newUpload.Key = s3EncodeName(upload.Object, encodingType)
-		newUpload.Initiated = upload.Initiated.UTC().Format(timeFormatAMZLong)
+		newUpload.Initiated = upload.Initiated.UTC().Format(iso8601TimeFormat)
 		listMultipartUploadsResponse.Uploads[index] = newUpload
 	}
 	return listMultipartUploadsResponse

--- a/cmd/bucket-quota.go
+++ b/cmd/bucket-quota.go
@@ -197,7 +197,7 @@ func enforceFIFOQuota(ctx context.Context, objectAPI ObjectLayer) error {
 		numKeys := len(scorer.fileNames())
 		for i, key := range scorer.fileNames() {
 			objects = append(objects, key)
-			if len(objects) < maxObjectList && (i < numKeys-1) {
+			if len(objects) < maxDeleteList && (i < numKeys-1) {
 				// skip deletion until maxObjectList or end of slice
 				continue
 			}

--- a/cmd/daily-lifecycle-ops.go
+++ b/cmd/daily-lifecycle-ops.go
@@ -85,7 +85,7 @@ func lifecycleRound(ctx context.Context, objAPI ObjectLayer) error {
 		for {
 			var objects []string
 			for obj := range objInfoCh {
-				if len(objects) == maxObjectList {
+				if len(objects) == maxDeleteList {
 					// Reached maximum delete requests, attempt a delete for now.
 					break
 				}

--- a/cmd/disk-cache.go
+++ b/cmd/disk-cache.go
@@ -99,14 +99,14 @@ func (c *cacheObjects) updateMetadataIfChanged(ctx context.Context, dcache *disk
 	bkMeta := make(map[string]string)
 	cacheMeta := make(map[string]string)
 	for k, v := range bkObjectInfo.UserDefined {
-		if HasPrefix(k, ReservedMetadataPrefix) {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) {
 			// Do not need to send any internal metadata
 			continue
 		}
 		bkMeta[http.CanonicalHeaderKey(k)] = v
 	}
 	for k, v := range cacheObjInfo.UserDefined {
-		if HasPrefix(k, ReservedMetadataPrefix) {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) {
 			// Do not need to send any internal metadata
 			continue
 		}

--- a/cmd/handler-utils.go
+++ b/cmd/handler-utils.go
@@ -104,6 +104,8 @@ func isDirectiveReplace(value string) bool {
 var userMetadataKeyPrefixes = []string{
 	"X-Amz-Meta-",
 	"X-Minio-Meta-",
+	"x-amz-meta-",
+	"x-minio-meta-",
 }
 
 // extractMetadata extracts metadata from HTTP header and HTTP queryString.

--- a/cmd/http/headers.go
+++ b/cmd/http/headers.go
@@ -58,7 +58,7 @@ const (
 
 	// S3 object tagging
 	AmzObjectTagging = "X-Amz-Tagging"
-	AmzTagCount      = "X-Amz-Tagging-Count"
+	AmzTagCount      = "x-amz-tagging-count"
 	AmzTagDirective  = "X-Amz-Tagging-Directive"
 
 	// S3 extensions

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -1072,7 +1072,6 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	// If x-amz-tagging-directive header is REPLACE, get passed tags.
 	if isDirectiveReplace(r.Header.Get(xhttp.AmzTagDirective)) {
 		objTags = r.Header.Get(xhttp.AmzObjectTagging)
-		srcInfo.UserDefined[xhttp.AmzTagDirective] = replaceDirective
 		if _, err := tags.ParseObjectTags(objTags); err != nil {
 			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 			return
@@ -1096,7 +1095,7 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	retentionMode, retentionDate, legalHold, s3Err := checkPutObjectLockAllowed(ctx, r, dstBucket, dstObject, getObjectInfo, retPerms, holdPerms)
 	if s3Err == ErrNone && retentionMode.Valid() {
 		srcInfo.UserDefined[strings.ToLower(xhttp.AmzObjectLockMode)] = string(retentionMode)
-		srcInfo.UserDefined[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(time.RFC3339)
+		srcInfo.UserDefined[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(iso8601TimeFormat)
 	}
 	if s3Err == ErrNone && legalHold.Status.Valid() {
 		srcInfo.UserDefined[strings.ToLower(xhttp.AmzObjectLockLegalHold)] = string(legalHold.Status)
@@ -1429,7 +1428,7 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	retentionMode, retentionDate, legalHold, s3Err := checkPutObjectLockAllowed(ctx, r, bucket, object, getObjectInfo, retPerms, holdPerms)
 	if s3Err == ErrNone && retentionMode.Valid() {
 		metadata[strings.ToLower(xhttp.AmzObjectLockMode)] = string(retentionMode)
-		metadata[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(time.RFC3339)
+		metadata[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(iso8601TimeFormat)
 	}
 	if s3Err == ErrNone && legalHold.Status.Valid() {
 		metadata[strings.ToLower(xhttp.AmzObjectLockLegalHold)] = string(legalHold.Status)
@@ -1605,7 +1604,7 @@ func (api objectAPIHandlers) NewMultipartUploadHandler(w http.ResponseWriter, r 
 	retentionMode, retentionDate, legalHold, s3Err := checkPutObjectLockAllowed(ctx, r, bucket, object, getObjectInfo, retPerms, holdPerms)
 	if s3Err == ErrNone && retentionMode.Valid() {
 		metadata[strings.ToLower(xhttp.AmzObjectLockMode)] = string(retentionMode)
-		metadata[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(time.RFC3339)
+		metadata[strings.ToLower(xhttp.AmzObjectLockRetainUntilDate)] = retentionDate.UTC().Format(iso8601TimeFormat)
 	}
 	if s3Err == ErrNone && legalHold.Status.Valid() {
 		metadata[strings.ToLower(xhttp.AmzObjectLockLegalHold)] = string(legalHold.Status)

--- a/cmd/post-policy_test.go
+++ b/cmd/post-policy_test.go
@@ -33,14 +33,13 @@ import (
 )
 
 const (
-	expirationDateFormat = "2006-01-02T15:04:05.999Z"
-	iso8601DateFormat    = "20060102T150405Z"
+	iso8601DateFormat = "20060102T150405Z"
 )
 
 func newPostPolicyBytesV4WithContentRange(credential, bucketName, objectKey string, expiration time.Time) []byte {
 	t := UTCNow()
 	// Add the expiration date.
-	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(expirationDateFormat))
+	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(iso8601TimeFormat))
 	// Add the bucket condition, only accept buckets equal to the one passed.
 	bucketConditionStr := fmt.Sprintf(`["eq", "$bucket", "%s"]`, bucketName)
 	// Add the key condition, only accept keys equal to the one passed.
@@ -71,7 +70,7 @@ func newPostPolicyBytesV4WithContentRange(credential, bucketName, objectKey stri
 func newPostPolicyBytesV4(credential, bucketName, objectKey string, expiration time.Time) []byte {
 	t := UTCNow()
 	// Add the expiration date.
-	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(expirationDateFormat))
+	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(iso8601TimeFormat))
 	// Add the bucket condition, only accept buckets equal to the one passed.
 	bucketConditionStr := fmt.Sprintf(`["eq", "$bucket", "%s"]`, bucketName)
 	// Add the key condition, only accept keys equal to the one passed.
@@ -98,7 +97,7 @@ func newPostPolicyBytesV4(credential, bucketName, objectKey string, expiration t
 // newPostPolicyBytesV2 - creates a bare bones postpolicy string with key and bucket matches.
 func newPostPolicyBytesV2(bucketName, objectKey string, expiration time.Time) []byte {
 	// Add the expiration date.
-	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(expirationDateFormat))
+	expirationStr := fmt.Sprintf(`"expiration": "%s"`, expiration.Format(iso8601TimeFormat))
 	// Add the bucket condition, only accept buckets equal to the one passed.
 	bucketConditionStr := fmt.Sprintf(`["eq", "$bucket", "%s"]`, bucketName)
 	// Add the key condition, only accept keys equal to the one passed.
@@ -264,7 +263,7 @@ func testPostPolicyBucketHandler(obj ObjectLayer, instanceType string, t TestErr
 			expectedRespStatus: http.StatusNoContent,
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			dates:              []interface{}{curTimePlus5Min.Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
+			dates:              []interface{}{curTimePlus5Min.Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
 			policy:             `{"expiration": "%s","conditions":[["eq", "$bucket", "` + bucketName + `"], ["starts-with", "$key", "test/"], ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"], ["eq", "$x-amz-date", "%s"], ["eq", "$x-amz-credential", "` + credentials.AccessKey + `/%s/us-east-1/s3/aws4_request"],["eq", "$x-amz-meta-uuid", "1234"]]}`,
 		},
 		// Corrupted Base 64 result
@@ -274,7 +273,7 @@ func testPostPolicyBucketHandler(obj ObjectLayer, instanceType string, t TestErr
 			expectedRespStatus: http.StatusBadRequest,
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			dates:              []interface{}{curTimePlus5Min.Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
+			dates:              []interface{}{curTimePlus5Min.Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
 			policy:             `{"expiration": "%s","conditions":[["eq", "$bucket", "` + bucketName + `"], ["starts-with", "$key", "test/"], ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"], ["eq", "$x-amz-date", "%s"], ["eq", "$x-amz-credential", "` + credentials.AccessKey + `/%s/us-east-1/s3/aws4_request"]]}`,
 			corruptedBase64:    true,
 		},
@@ -285,7 +284,7 @@ func testPostPolicyBucketHandler(obj ObjectLayer, instanceType string, t TestErr
 			expectedRespStatus: http.StatusBadRequest,
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			dates:              []interface{}{curTimePlus5Min.Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
+			dates:              []interface{}{curTimePlus5Min.Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
 			policy:             `{"expiration": "%s","conditions":[["eq", "$bucket", "` + bucketName + `"], ["starts-with", "$key", "test/"], ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"], ["eq", "$x-amz-date", "%s"], ["eq", "$x-amz-credential", "` + credentials.AccessKey + `/%s/us-east-1/s3/aws4_request"]]}`,
 			corruptedMultipart: true,
 		},
@@ -307,7 +306,7 @@ func testPostPolicyBucketHandler(obj ObjectLayer, instanceType string, t TestErr
 			expectedRespStatus: http.StatusForbidden,
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			dates:              []interface{}{curTime.Add(-1 * time.Minute * 5).Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
+			dates:              []interface{}{curTime.Add(-1 * time.Minute * 5).Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
 			policy:             `{"expiration": "%s","conditions":[["eq", "$bucket", "` + bucketName + `"], ["starts-with", "$key", "test/"], ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"], ["eq", "$x-amz-date", "%s"], ["eq", "$x-amz-credential", "` + credentials.AccessKey + `/%s/us-east-1/s3/aws4_request"]]}`,
 		},
 		// Corrupted policy document
@@ -317,7 +316,7 @@ func testPostPolicyBucketHandler(obj ObjectLayer, instanceType string, t TestErr
 			expectedRespStatus: http.StatusForbidden,
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			dates:              []interface{}{curTimePlus5Min.Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
+			dates:              []interface{}{curTimePlus5Min.Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)},
 			policy:             `{"3/aws4_request"]]}`,
 		},
 	}
@@ -460,7 +459,7 @@ func testPostPolicyBucketHandlerRedirect(obj ObjectLayer, instanceType string, t
 	// initialize HTTP NewRecorder, this records any mutations to response writer inside the handler.
 	rec := httptest.NewRecorder()
 
-	dates := []interface{}{curTimePlus5Min.Format(expirationDateFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)}
+	dates := []interface{}{curTimePlus5Min.Format(iso8601TimeFormat), curTime.Format(iso8601DateFormat), curTime.Format(yyyymmdd)}
 	policy := `{"expiration": "%s","conditions":[["eq", "$bucket", "` + bucketName + `"], {"success_action_redirect":"` + redirectURL.String() + `"},["starts-with", "$key", "test/"], ["eq", "$x-amz-meta-uuid", "1234"], ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"], ["eq", "$x-amz-date", "%s"], ["eq", "$x-amz-credential", "` + credentials.AccessKey + `/%s/us-east-1/s3/aws4_request"]]}`
 
 	// Generate the final policy document

--- a/cmd/server_test.go
+++ b/cmd/server_test.go
@@ -1236,7 +1236,7 @@ func (s *TestSuiteCommon) TestListBuckets(c *check) {
 	c.Assert(createdBucket.Name != "", true)
 
 	// Parse the bucket modtime
-	creationTime, err := time.Parse(timeFormatAMZLong, createdBucket.CreationDate)
+	creationTime, err := time.Parse(iso8601TimeFormat, createdBucket.CreationDate)
 	c.Assert(err, nil)
 
 	// Check if bucket modtime is consistent (not less than current time and not late more than 5 minutes)

--- a/cmd/web-handlers.go
+++ b/cmd/web-handlers.go
@@ -409,8 +409,8 @@ func (web *webAPIHandlers) ListObjects(r *http.Request, args *ListObjectsArgs, r
 		nextMarker := ""
 		// Fetch all the objects
 		for {
-			result, err := core.ListObjects(args.BucketName, args.Prefix, nextMarker, SlashSeparator,
-				maxObjectList)
+			// Let listObjects reply back the maximum from server implementation
+			result, err := core.ListObjects(args.BucketName, args.Prefix, nextMarker, SlashSeparator, 0)
 			if err != nil {
 				return toJSONError(ctx, err, args.BucketName)
 			}
@@ -524,7 +524,8 @@ func (web *webAPIHandlers) ListObjects(r *http.Request, args *ListObjectsArgs, r
 	nextMarker := ""
 	// Fetch all the objects
 	for {
-		lo, err := listObjects(ctx, args.BucketName, args.Prefix, nextMarker, SlashSeparator, maxObjectList)
+		// Limit browser to 1000 batches to be more responsive, scrolling friendly.
+		lo, err := listObjects(ctx, args.BucketName, args.Prefix, nextMarker, SlashSeparator, 1000)
 		if err != nil {
 			return &json2.Error{Message: err.Error()}
 		}
@@ -761,7 +762,7 @@ next:
 		for {
 			var objects []string
 			for obj := range objInfoCh {
-				if len(objects) == maxObjectList {
+				if len(objects) == maxDeleteList {
 					// Reached maximum delete requests, attempt a delete for now.
 					break
 				}
@@ -1122,7 +1123,7 @@ func (web *webAPIHandlers) Upload(w http.ResponseWriter, r *http.Request) {
 		retentionMode, retentionDate, legalHold, s3Err := checkPutObjectLockAllowed(ctx, r, bucket, object, getObjectInfo, retPerms, holdPerms)
 		if s3Err == ErrNone && retentionMode != "" {
 			opts.UserDefined[xhttp.AmzObjectLockMode] = string(retentionMode)
-			opts.UserDefined[xhttp.AmzObjectLockRetainUntilDate] = retentionDate.UTC().Format(time.RFC3339)
+			opts.UserDefined[xhttp.AmzObjectLockRetainUntilDate] = retentionDate.UTC().Format(iso8601TimeFormat)
 		}
 		if s3Err == ErrNone && legalHold.Status != "" {
 			opts.UserDefined[xhttp.AmzObjectLockLegalHold] = string(legalHold.Status)

--- a/cmd/xl-zones.go
+++ b/cmd/xl-zones.go
@@ -1046,7 +1046,7 @@ func (z *xlZones) PutObjectPart(ctx context.Context, bucket, object, uploadID st
 		return z.zones[0].PutObjectPart(ctx, bucket, object, uploadID, partID, data, opts)
 	}
 	for _, zone := range z.zones {
-		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxObjectList)
+		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxUploadsList)
 		if err != nil {
 			return PartInfo{}, err
 		}
@@ -1078,7 +1078,7 @@ func (z *xlZones) ListObjectParts(ctx context.Context, bucket, object, uploadID 
 		return z.zones[0].ListObjectParts(ctx, bucket, object, uploadID, partNumberMarker, maxParts, opts)
 	}
 	for _, zone := range z.zones {
-		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxObjectList)
+		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxUploadsList)
 		if err != nil {
 			return ListPartsInfo{}, err
 		}
@@ -1110,7 +1110,7 @@ func (z *xlZones) AbortMultipartUpload(ctx context.Context, bucket, object, uplo
 	}
 
 	for _, zone := range z.zones {
-		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxObjectList)
+		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxUploadsList)
 		if err != nil {
 			return err
 		}
@@ -1157,7 +1157,7 @@ func (z *xlZones) CompleteMultipartUpload(ctx context.Context, bucket, object, u
 	}
 
 	for _, zone := range z.zones {
-		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxObjectList)
+		result, err := zone.ListMultipartUploads(ctx, bucket, object, "", "", "", maxUploadsList)
 		if err != nil {
 			return objInfo, err
 		}

--- a/mint/run/core/aws-sdk-php/quick-tests.php
+++ b/mint/run/core/aws-sdk-php/quick-tests.php
@@ -33,7 +33,7 @@ const HTTP_NOCONTENT = "204";
 const HTTP_BADREQUEST = "400";
 const HTTP_NOTIMPLEMENTED = "501";
 const HTTP_INTERNAL_ERROR = "500";
-const TEST_METADATA = ['Param_1' => 'val-1'];
+const TEST_METADATA = ['param_1' => 'val-1'];
 
 /**
  * ClientConfig abstracts configuration details to connect to a

--- a/pkg/event/event.go
+++ b/pkg/event/event.go
@@ -24,7 +24,7 @@ const (
 	AccessFormat = "access"
 
 	// AMZTimeFormat - event time format.
-	AMZTimeFormat = "2006-01-02T15:04:05Z"
+	AMZTimeFormat = "2006-01-02T15:04:05.000Z"
 )
 
 // Identity represents access key who caused the event.

